File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.java
import java.util.*;

class Solution {
    static class Node {
        int x, y, id;
        Node left, right;

        Node(int x, int y, int id) {
            this.x = x;
            this.y = y;
            this.id = id;
        }
    }

    private static void insert(Node root, Node child) {
        Node cur = root;
        while (true) {
            if (child.x < cur.x) {
                if (cur.left == null) {
                    cur.left = child;
                    return;
                }
                cur = cur.left;
            } else {
                if (cur.right == null) {
                    cur.right = child;
                    return;
                }
                cur = cur.right;
            }
        }
    }

    private static void preorder(Node node, List<Integer> out) {
        if (node == null) return;
        out.add(node.id);
        preorder(node.left, out);
        preorder(node.right, out);
    }

    private static void postorder(Node node, List<Integer> out) {
        if (node == null) return;
        postorder(node.left, out);
        postorder(node.right, out);
        out.add(node.id);
    }

    public int[][] solution(int[][] nodeinfo) {
        int n = nodeinfo.length;

        int[][] nodes = new int[n][3]; // [x, y, id]
        for (int i = 0; i < n; i++) {
            nodes[i][0] = nodeinfo[i][0];
            nodes[i][1] = nodeinfo[i][1];
            nodes[i][2] = i + 1;
        }

        Arrays.sort(nodes, (a, b) -> {
            if (a[1] != b[1]) return b[1] - a[1]; // y desc
            return a[0] - b[0];                   // x asc
        });

        Node root = new Node(nodes[0][0], nodes[0][1], nodes[0][2]);
        for (int i = 1; i < n; i++) {
            insert(root, new Node(nodes[i][0], nodes[i][1], nodes[i][2]));
        }

        List<Integer> pre = new ArrayList<>(n);
        List<Integer> post = new ArrayList<>(n);

        preorder(root, pre);
        postorder(root, post);

        int[][] answer = new int[2][n];
        for (int i = 0; i < n; i++) {
            answer[0][i] = pre.get(i);
            answer[1][i] = post.get(i);
        }
        return answer;
    }
}

File: 프로그래머스/3/42892. 길 찾기 게임/길 찾기 게임.kt
class Solution {

    private class Node(val x: Int, val y: Int, val id: Int) {
        var left: Node? = null
        var right: Node? = null
    }

    private fun insert(root: Node, child: Node) {
        var cur: Node? = root
        while (cur != null) {
            if (child.x < cur.x) {
                if (cur.left == null) {
                    cur.left = child
                    return
                }
                cur = cur.left
            } else {
                if (cur.right == null) {
                    cur.right = child
                    return
                }
                cur = cur.right
            }
        }
    }

    private fun preorder(node: Node?, out: MutableList<Int>) {
        if (node == null) return
        out.add(node.id)
        preorder(node.left, out)
        preorder(node.right, out)
    }

    private fun postorder(node: Node?, out: MutableList<Int>) {
        if (node == null) return
        postorder(node.left, out)
        postorder(node.right, out)
        out.add(node.id)
    }

    fun solution(nodeinfo: Array<IntArray>): Array<IntArray> {
        val n = nodeinfo.size

        // nodes: intArrayOf(x, y, id)
        val nodes = ArrayList<IntArray>(n)
        for (i in 0 until n) {
            nodes.add(intArrayOf(nodeinfo[i][0], nodeinfo[i][1], i + 1))
        }

        // y desc, x asc
        nodes.sortWith(compareByDescending<IntArray> { it[1] }.thenBy { it[0] })

        // build tree
        val root = Node(nodes[0][0], nodes[0][1], nodes[0][2])
        for (i in 1 until n) {
            insert(root, Node(nodes[i][0], nodes[i][1], nodes[i][2]))
        }

        val pre = ArrayList<Int>(n)
        val post = ArrayList<Int>(n)

        preorder(root, pre)
        postorder(root, post)

        val answer = Array(2) { IntArray(n) }
        for (i in 0 until n) {
            answer[0][i] = pre[i]
            answer[1][i] = post[i]
        }
        return answer
    }
}